src/net.cpp

```diff
@@ -1122,32 +1122,36 @@ void CConnman::DisconnectNodes()
 
         // Disconnect unused nodes
         std::vector<CNode*> nodes_copy = m_nodes;
-        for (CNode* pnode : nodes_copy)
         {
-            if (pnode->fDisconnect)
-            {
-                // remove from m_nodes
-                m_nodes.erase(remove(m_nodes.begin(), m_nodes.end(), pnode), m_nodes.end());
+            LOCK(m_disconnect_mutex);
+            for (CNode* pnode : nodes_copy) {
+                if (pnode->fDisconnect) {
+                    // remove from m_nodes
+                    m_nodes.erase(remove(m_nodes.begin(), m_nodes.end(), pnode), m_nodes.end());
 
-                // release outbound grant (if any)
-                pnode->grantOutbound.Release();
+                    // release outbound grant (if any)
+                    pnode->grantOutbound.Release();
 
-                // close socket and cleanup
-                pnode->CloseSocketDisconnect();
+                    // close socket and cleanup
+                    pnode->CloseSocketDisconnect();
 
-                // hold in disconnected pool until all refs are released
-                pnode->Release();
-                m_nodes_disconnected.push_back(pnode);
+                    // hold in disconnected pool until all refs are released
+                    pnode->Release();
+                    m_nodes_disconnected.push_back(pnode);
+                }
             }
         }
     }
     {
+        LOCK(m_disconnect_mutex);
         // Delete disconnected nodes
         std::list<CNode*> nodes_disconnected_copy = m_nodes_disconnected;
         for (CNode* pnode : nodes_disconnected_copy)
         {
             // Destroy the object only after other threads have stopped using it.
-            if (pnode->GetRefCount() <= 0) {
+            // Prevent two threads trying to delete the same node: set nRefCount to -1 first
```

Comment:
I think `m_disconnect_mutex` already prevents this from happening

Reply:
Hmm you may be right, I might have gone too far with that. What I was trying to protect against was checking the number of references in one instruction, and permitting another thread to increase the `refCount` in the meantime before we remove the node in the subsequent lines. But, it looks like we don't have any functions which would increase the `refCount` during normal operation (only incremented on new connections), so probably unnecessary here.

If I'm honest, I kind of prefer this belt-and-suspenders way myself, but would be easily persuaded that it's superfluous or worse than current behaviour (useless extra locking).

I felt a bit nevous about nesting `m_disconnect_mutex` inside of `m_nodes_mutex`, but it seemed like it was slightly preferable to just locking `m_nodes_mutex` for the whole of `DisconnectNodes()` as that's used in many other operations. I don't think there is any way `StopNodes()` and `DisconnectNodes()` can lock each other here though. Would you agree?

Reply:
The CNode reference counting is a total nightmare and we should (and have in the past: https://github.com/bitcoin/bitcoin/pull/10738/) try to get rid of it at some point.

> permitting another thread to increase the refCount in the meantime before we remove the node in the subsequent lines.

This is actually possible with and without your patch and the way we protect against this is by not calling `CNode::AddRef` after nodes where moved to the disconnection queue.

```c++
if (pnode->nRefCount.compare_exchange_strong(expectedRefCount, -1)) {
    // Theoretically, nothing stops a different thread from calling `AddRef()` while we are here
    m_nodes_disconnected.remove(pnode);
    DeleteNode(pnode);
}
```

>  But, it looks like we don't have any functions which would increase the refCount during normal operation (only incremented on new connections), so probably unnecessary here.

Our RAII helper [`NodesSnapshot`](https://github.com/bitcoin/bitcoin/blob/7f0b79ea132d22ad5212c1d3ff4325715ca5ac12/src/net.h#L1189) does call `AddRef` during normal operation (e.g. before processing messages) but `NodesSnapshot` doesn't take a snapshot of `m_nodes_disconnected` so it's fine.

> I don't think there is any way StopNodes() and DisconnectNodes() can lock each other here though. 

Agreed, as long as you always lock in the same order (which afaict you do).